Sources/SwiftInspectorVisitors/ExtensionVisitor.swift

```diff
@@ -0,0 +1,120 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import SwiftSyntax
+
+public final class ExtensionVisitor: SyntaxVisitor {
+
+  /// The extension found by this visitor.
+  public private(set) var extensionInfo: ExtensionInfo?
+  public private(set) var structs = [StructInfo]()
+
+  // TODO: also find and nested classes
+  // TODO: also find and nested enums
+
+  public override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
+
+    guard !hasFinishedParsingExtension else {
+      assertionFailure("Encountered more than one top-level extension. This is a usage error: a single ExtensionVisitor instance should start walking only over a node of type `ExtensionDeclSyntax`")
+      return .skipChildren
+    }
+
+    // There is no known case where these identifiers can not be cast to SimpleTypeIdentifierSyntax.
+    let name = node.extendedType.as(SimpleTypeIdentifierSyntax.self)!.name.text
+    let typeInheritanceVisitor = TypeInheritanceVisitor()
+    typeInheritanceVisitor.walk(node)
+    let genericRequirementsVisitor = GenericRequirementVisitor()
+    genericRequirementsVisitor.walk(node)
+
+    extensionInfo = ExtensionInfo(
+      name: name,
+      inheritsFromTypes: typeInheritanceVisitor.inheritsFromTypes,
+      genericRequirements: genericRequirementsVisitor.genericRequirements,
+      innerStructs: [])
+    return .visitChildren
+  }
+
+  public override func visitPost(_ node: ExtensionDeclSyntax) {
+    hasFinishedParsingExtension = true
+  }
+
+  public override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+    if !hasFinishedParsingExtension, let _ = extensionInfo {
+      // We've previously found an extension declaration, so this must be an inner class.
+      // TODO: Utilize class visitor to find inner class information, utilizing parent information from extensionInfo
+    } else {
+      // We've encountered a class declaration before encountering an extension declaration. Something is wrong.
+      assertionFailure("Encountered a top-level class. This is a usage error: a single ExtensionVisitor instance should start walking only over a node of type `ExtensionDeclSyntax`")
+    }
+    return .skipChildren
+  }
+
+  public override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+    if !hasFinishedParsingExtension, let _ = extensionInfo {
+      // We've previously found an extension declaration, so this must be an inner struct.
+      let structVisitor = StructVisitor(parentTypeName: extensionInfo?.name)
```

Comment:
Why the optionality if we can pass a non-optional?

Reply:
😂 good catch

Reply:
d237eff